subsys/fs/fs.c

```diff
@@ -0,0 +1,529 @@
+		rc = zfp->fops->sync(zfp);
+		if (rc < 0) {
+			SYS_LOG_ERR("file sync error (%d)\n", rc);
+		}
+	}
+
+	return rc;
+}
+
+/* Directory operations */
+int fs_opendir(fs_dir_t *zdp, const char *abs_path)
+{
+	fs_mount_t *mp;
+	const char *fs_path;
+	int rc, match_len;
+
+	if ((strlen(abs_path) <= 1) || (abs_path[0] != '/')) {
+		SYS_LOG_ERR("invalid file name!!\n");
+		return -EINVAL;
+	}
+
+	rc = get_mnt_point(&mp, abs_path, &match_len);
+	if (rc < 0) {
+		SYS_LOG_ERR("%s:mount point not found!!\n", __func__);
+		return rc;
+	}
+	fs_path = &abs_path[match_len];
+
+	zdp->mp = mp;
+	zdp->dops = mp->fs->dops;
+
+	if (zdp->dops->opendir != NULL) {
+		rc = zdp->dops->opendir(zdp, fs_path);
+		if (rc < 0) {
+			SYS_LOG_ERR("directory open error (%d)\n", rc);
+		}
+	}
+
+	return rc;
+}
+
+int fs_readdir(fs_dir_t *zdp, struct fs_dirent *entry)
+{
+	int rc = -EINVAL;
+
+	if (zdp->dops->readdir != NULL) {
+		rc = zdp->dops->readdir(zdp, entry);
+		if (rc < 0) {
+			SYS_LOG_ERR("directory read error (%d)\n", rc);
+		}
+	}
+	return rc;
+}
+
+int fs_closedir(fs_dir_t *zdp)
+{
+	int rc = -EINVAL;
+
+	if (zdp->dops->closedir != NULL) {
+		rc = zdp->dops->closedir(zdp);
+		if (rc < 0) {
+			SYS_LOG_ERR("directory close error (%d)\n", rc);
+			return rc;
+		}
+	}
+
+	zdp->mp = NULL;
+	zdp->dops = NULL;
+	return rc;
+}
+
+/* Filesystem operations */
+int fs_mkdir(const char *abs_path)
+{
+	fs_mount_t *mp;
+	const char *fs_path;
+	int rc, match_len;
+
+	if ((strlen(abs_path) <= 1) || (abs_path[0] != '/')) {
+		SYS_LOG_ERR("invalid file name!!\n");
+		return -EINVAL;
+	}
+
+	rc = get_mnt_point(&mp, abs_path, &match_len);
+	if (rc < 0) {
+		SYS_LOG_ERR("%s:mount point not found!!\n", __func__);
+		return rc;
+	}
+	fs_path = &abs_path[match_len];
+
+	if (mp->fs->fs_ops->mkdir != NULL) {
+		rc = mp->fs->fs_ops->mkdir(mp, fs_path);
+		if (rc < 0) {
+			SYS_LOG_ERR("failed to create directory (%d)\n", rc);
+		}
+	}
+
+	return rc;
+}
+
+int fs_unlink(const char *abs_path)
+{
+	fs_mount_t *mp;
+	const char *fs_path;
+	int rc, match_len;
+
+	if ((strlen(abs_path) <= 1) || (abs_path[0] != '/')) {
+		SYS_LOG_ERR("invalid file name!!\n");
+		return -EINVAL;
+	}
+
+	rc = get_mnt_point(&mp, abs_path, &match_len);
+	if (rc < 0) {
+		SYS_LOG_ERR("%s:mount point not found!!\n", __func__);
+		return rc;
+	}
+	fs_path = &abs_path[match_len];
+
+	if (mp->fs->fs_ops->unlink != NULL) {
+		rc = mp->fs->fs_ops->unlink(mp, fs_path);
+		if (rc < 0) {
+			SYS_LOG_ERR("failed to unlink path (%d)\n", rc);
+		}
+	}
+
+	return rc;
+}
+
+int fs_stat(const char *abs_path, struct fs_dirent *entry)
+{
+	fs_mount_t *mp;
+	const char *fs_path;
+	int rc, match_len;
+
+	if ((strlen(abs_path) <= 1) || (abs_path[0] != '/')) {
+		SYS_LOG_ERR("invalid file name!!\n");
+		return -EINVAL;
+	}
+
+	rc = get_mnt_point(&mp, abs_path, &match_len);
+	if (rc < 0) {
+		SYS_LOG_ERR("%s:mount point not found!!\n", __func__);
+		return rc;
+	}
+	fs_path = &abs_path[match_len];
+
+	if (mp->fs->fs_ops->stat != NULL) {
```

Comment:
Suggestion: this indirection can be reduced if `mp->fs` points to a structure with all the function pointers thrown in instead of having 3 separate structs for file, filesystem, and directory operations:

```C
if (mp->fs->stat != NULL) {
    ...
}
```

Reply:
Ok. In that case, single struct contains all the functions pointers to file, directory and file system operations.